_posts/2019-01-XX-meltdown-and-spectre-vulnerabilities.md

```diff
@@ -0,0 +1,150 @@
+what they are and how they are threatening the users of modern machines.
+
+# Spectre
+
+![Spectre](https://3mdeb.com/wp-content/uploads/2017/07/spectre.png)
+
+Spectre refers to a vulnerability which breaks the isolation between different
+applications. It allows an attacker to trick error-free programs, which follow
+best practices, into leaking their secrets.
+
+It is basically divided into few variants:
+
+- [CVE-2017-5753] Spectre Variant 1 - bounds check bypass.
+- [CVE-2017-5715] Spectre Variant 2 - branch target injection
+- [CVE-2017-5754] Variant 3, Meltdown - rogue data cache load
+- [CVE-2018-3640] Variant 3a - rogue system register read
+- [CVE-2018-3639] Variant 4 - speculative store bypass
+
+## Spectre Variant 1 - bounds check bypass
+
+Variant 1 relies on misleading the processor to do illegal actions. How was it
+achieved?
+
+On the initial phase of the attack, the processor is trained with valid inputs
+in arbitrary attack program. Basically, we prepare processor to be cheated.
+Let's say it is an if statement which during training phase evaluates to true:
+
+```
+if (x < array1_size)
+    y = array2[array1[x] * 4096];
+```
+
+Next, during the exploit phase, the attacker invokes the code with a value of
+`x` outside the bounds of `array1`. What happens now?
+
+Due to the training phase, which task was to teach the branch predictor (which
+is the source of the vulnerability variant) that the if statement will be true,
+the processor will think that the if statement still will be true (what?!).
+
+> The branch means execution flow. If we have if statement, we divide the
+> execution flow into two branches: condition is true or condition is false.
+
+That is how speculative execution works. Rather than waiting for determination
+of the branch result, the CPU guesses that the bounds check will be true and
+already speculatively executes instructions that evaluate
+`array2[array1[x]*4096]` using the malicious `x`. Given that, the process
+accesses data it is not permitted to. `x` is chosen by the attacker and the
+read from `array2` is dependent on `arrray1[x]`. Even if the CPU discovers
+its error and reverts the changes, the secret data still exists in the cache
+and can be retrieved by analyzing the cache by an attacker.
+
+## Spectre Variant 2 - branch target injection
+
+Variant 2 is very similar to the first variant. In the previous example, the
+malicious code resided in attackers address space. While in the second variant,
+the attacker chooses a specific "gadget" from victim's address space. The
+processor is trained to mispredict a branch and to execute the "gadget" lying
+in the victim's address space. Information revealed by the "gadget" remains in
+the cache like before and the attacker can analyze it to retrieve the
+information.
+
+## Variant 3 - Meltdown, rogue data cache load
+
+![Meltdown](https://3mdeb.com/wp-content/uploads/2017/07/meltdown.png)
```

Comment:
Same here, CC0

Reply:
@miczyg1 that means we definitely can use that.